tensorcircuit/templates/lattice.py

```diff
@@ -0,0 +1,1555 @@
+            ref_dist_matrix_sq.flatten(), max_k, tol
+        )
+        if not dist_shells_sq:
+            return
+
+        # 3. Create "neighbor templates" based on identifier displacements
+        neighbor_templates: Dict[int, Dict[int, List[Tuple[Coordinates, int]]]] = {
+            b: {k: [] for k in range(1, len(dist_shells_sq) + 1)}
+            for b in range(self.num_basis)
+        }
+
+        for k_idx, target_d_sq in enumerate(dist_shells_sq):
+            k = k_idx + 1
+            # Find which pairs (ref_basis_site, any_site) match this distance
+            match_indices = np.where(
+                np.isclose(ref_dist_matrix_sq, target_d_sq, rtol=0, atol=tol**2)
+            )
+            # match_indices is a tuple of (array_of_basis_indices, array_of_site_indices)
+            for source_basis_idx, target_site_idx in zip(*match_indices):
+                source_ident = self._identifiers[source_basis_idx]  # (0,..,0,basis)
+                target_ident = self._identifiers[target_site_idx]
+
+                # Rule: (uc_displacement, target_basis_index)
+                uc_disp = np.array(target_ident[:-1]) - np.array(source_ident[:-1])
+                target_basis = target_ident[-1]
+                neighbor_templates[source_basis_idx][k].append((uc_disp, target_basis))
+
+        # 4. Apply the templates to all sites in the lattice
+        self._neighbor_maps = {k: {} for k in range(1, len(dist_shells_sq) + 1)}
+        size_arr = np.array(self.size)
+
+        for i in range(self.num_sites):
+            source_ident = self._identifiers[i]
+            source_uc = np.array(source_ident[:-1])
+            source_basis = source_ident[-1]
+
+            for k, templates in neighbor_templates[source_basis].items():
+                neighbors_of_i_for_k = []
+                for uc_disp, target_basis in templates:
+                    target_uc = source_uc + uc_disp
+
+                    # Wrap target_uc according to PBC rules
+                    is_in_bounds = True
+                    for dim in range(self.dimensionality):
+                        if self.pbc[dim]:
+                            target_uc[dim] %= size_arr[dim]
+                        elif not (0 <= target_uc[dim] < size_arr[dim]):
+                            is_in_bounds = False
+                            break
+
+                    if is_in_bounds:
+                        target_ident = tuple(target_uc) + (target_basis,)
+                        # The identifier must exist in the map
+                        if target_ident in self._ident_to_idx:
+                            neighbor_idx = self._ident_to_idx[target_ident]
+                            neighbors_of_i_for_k.append(neighbor_idx)
+
+                if neighbors_of_i_for_k:
+                    self._neighbor_maps[k][i] = sorted(neighbors_of_i_for_k)
+        logger.info("Caching the full distance matrix with Minimum Image Convention...")
```

Comment:
this `distance_matrix` should be obtained more cleverly, otherwise it will ruin the advantage for an independent neighbor search logic for fully PBC cases, as you again recompute the full distance_matrix from the scratch